### ISSUE TITLE
Integrate model with Triton

### ISSUE THREAD

**Issue by jalola** [NVIDIA/retinanet-examples#207]
Hi,

I want to integrate the TensorRT model to Triton https://github.com/NVIDIA/triton-inference-server

If I understand the code correctly, there can be 2 ways:
1. Use the engine.plan built from engine.cpp
Pros:
- The model will have decode + nms plugin, so it is fast
Cons:
- Need to build the (decode + nms) custom_plugin.so library then run Triton with this custom plugin. Otherwise, when loading the model, it will show error:
`Serialization Error in load: 0 (Cannot deserialize plugin since corresponding IPluginCreator not found in Plugin Registry)`
I still don't know how to build this custom_plugin.so and load it. Do you have any idea to build it?
- Cannot use batch_size > 1 for now, if I want to use it, I need to modify the engine.cpp I guess

2. Don't use the decode + nms plugin for TensorRT any more:
Pros:
- Can use batch_size > 1 like Pytorch
- No need to build custom_plugin.so for Triton
Cons:
- Decode + NMS will be slower due to running on cpu

Do you have any recommendation on this? Which way is better or there may be a better way to integrate TensorRT model to Triton?

Thanks

**Comment by yashnv** [comment]
Hello, you can go to extras -> cppapi and follow build steps(`cmake` and `make -j`), which creates a `libretinanet.so` library.

We are working on updating plugins to IPluginV2DynamicExt for easing the change in batch size, and not requiring to change the code for that.

**Comment by jalola** [comment]
I tried it before and it worked.
However, I actually don't want to install opencv into Triton to load that `libretinanet.so` so I guess I need to modify `CMakeLists.txt` and build it again. I only need the (decode + nms) plugin and I am using TensorRT fp16.

It is so great that you are working on the dynamic batch_size > 1. Thanks a lot and looking forward to new update.

**Comment by hcx1231** [comment]
Plugin creator registration succeeded,and it showed 
"The TRT engine doesn't specify appropriate dimensions to support dynamic batching."
I changed batch size to 1 but it did not  work.
How to solve it?

**Comment by jalola** [comment]
How did you build the TRT engine?
How did you change batch_size to 1, as I know it is hard coded as 1

**Comment by hcx1231** [comment]
> How did you build the TRT engine?
> How did you change batch_size to 1, as I know it is hard coded as 1

Sorry I made a mistake.I just set the 'max_batch_size' to 1 in config.pbtxt.
I build the TRT engine just by 'odtk export xxx.pth xxx.plan'

**Comment by hcx1231** [comment]
It finally worked by editing the config.pbtxt file.Thank you

> How did you build the TRT engine?
> How did you change batch_size to 1, as I know it is hard coded as 1

It finally worked by editing the config.pbtxt file.Thank you!

**Comment by nrgsy** [comment]
@jalola @hcx1231 what ended up working for either of you? I am attempting to do the same thing (use tensorRT-optimized retinanet within tritonserver). I have exported the pytorch *.pth model to a tensorRT *.plan model. But when I try to use the model via tritonserver, I get the same serialization error as the original post due to the missing plugin.

Sounds like the solution was to build the `libretinanet.so` library and copy it somewhere within tritonserver? I also need to avoid adding any additional dependencies into the tritonserver docker container, so sounds like I will need to remove everything related to opencv from `extras/cppapi/CMakeLists.txt` before building. Would either of you happen to have a copy of the CMakelists.txt that worked for you?

Thanks in advance.

**Comment by jalola** [comment]
> Hello, you can go to extras -> cppapi and follow build steps(`cmake` and `make -j`), which creates a `libretinanet.so` library.
> 
> We are working on updating plugins to IPluginV2DynamicExt for easing the change in batch size, and not requiring to change the code for that.

Hi @nrgsy ,
I think you can build the lib.so with opencv first then try to run your model. See the above instruction how to build .so
After that, you can modify the code `extras -> cppapi` to remove opencv dependencies. I have not done this one yet actually.

**Comment by nrgsy** [comment]
What ended up working for me was editing `csrc/engine.cpp` to remove the int8 calibration stuff, which I wasn't using and was the only thing that really needed OpenCV (csrc/calibrator.h). Then I removed all OpenCV-related stuff + calibrator.h from the CMakelists.txt and built `libretinanet.so`. After that I used the -eLD_PRELOAD docker flag with the .so file when running Tritionserver, it accepted my TensorRT retinanet model, and I was able to run inferences.

Thanks @jalola!
